agent/php_api.c

```diff
@@ -1578,3 +1578,64 @@ PHP_FUNCTION(newrelic_get_trace_metadata) {
     nr_free(span_id);
   }
 }
+
+/*
+ * Purpose      Allows a caller to add a user id string to agent attributes in 
+                transaction event, transaction trace, error and span.
+ *
+ */
+#ifdef TAGS
+void zif_newrelic_set_user_id(void); /* ctags landing pad only */
+void newrelic_set_user_id(void);     /* ctags landing pad only */
+#endif
+PHP_FUNCTION(newrelic_set_user_id) {
+  zval* uuid_zv = NULL;
+  char* uuid_str = NULL;
+
+  NR_UNUSED_RETURN_VALUE_PTR;
+  NR_UNUSED_THIS_PTR;
+  NR_UNUSED_RETURN_VALUE_USED;
+
+  nr_php_api_add_supportability_metric("set_user_id" TSRMLS_CC);
+
+  if (nrunlikely(NULL == NRPRG(txn))) {
+    nrl_verbosedebug(NRL_API, "newrelic_set_user_id failure: txn is NULL.");
+    RETURN_FALSE;
+  }
+
+  if (1 != ZEND_NUM_ARGS()) {
+    nrl_warning(NRL_API,
+                "newrelic_set_user_id failure: invalid number of parameters.");
+    RETURN_FALSE;
+  }
+
+  if (FAILURE == zend_parse_parameters(ZEND_NUM_ARGS(), "z", &uuid_zv)) {
+    nrl_warning(NRL_API,
+                "newrelic_set_user_id failure: invalid argument passed.");
+    RETURN_FALSE;
+  }
+
+  if (!nr_php_is_zval_non_empty_string(uuid_zv)) {
+    nrl_warning(
+        NRL_API,
+        "newrelic_set_user_id failure: User ID must be a non-empty string.");
+    RETURN_FALSE;
+  }
+
+  if (NR_ATTRIBUTE_VALUE_LENGTH_LIMIT < Z_STRLEN_P(uuid_zv)) {
+    nrl_warning(NRL_API,
+                "newrelic_set_user_id_failure: invalid string length.");
+    RETURN_FALSE;
+  }
+
+  uuid_str = nr_strdup(Z_STRVAL_P(uuid_zv));
```

Comment:
It's probably safer to call nr_strdup with the length explicitly passed in `str = nr_strndup(Z_STRVAL_P(uuid_zv), Z_STRLEN_P(uuid_zv));` like was done [here](https://github.com/newrelic/newrelic-php-agent/blob/dev/agent/php_api.c#L582)
since we can't depend on zval strings to be guaranteed to be null terminated.

references:
https://www.phpinternalsbook.com/php7/internal_types/strings/zend_strings.html#the-zend-string-api

Reply:
Added in 18dd7acfa2344abb5b9d538a432acd2d33948513